nni/retiarii/oneshot/pytorch/sampling.py

```diff
@@ -0,0 +1,193 @@
+        self.grad_clip = grad_clip
+
+    def configure_architecture_optimizers(self):
+        return optim.Adam(self.controller.parameters(), lr=3.5e-4)
+
+    @property
+    def default_replace_dict(self):
+        return {
+            LayerChoice : PathSamplingLayerChoice,
+            InputChoice : PathSamplingInputChoice
+        }
+
+    def training_step(self, batch, batch_idx):
+        # grad manually
+        opts = self.optimizers()
+        if isinstance(opts,list):
+            # list of optimizers
+            # pylint: disable=unsubscriptable-object
+            arc_opt = opts[0]
+            w_opt = opts[1:]
+        else :
+            # no model optimizer
+            arc_opt = opts
+
+        # the batch is composed with x, y, b, where b denote if the data provided
+        # is from the training set
+        batch, source = batch
+
+        if source == 'train':
+            # step 1: train model params
+            self._resample()
+            self.call_user_optimizers(w_opt, 'zero_grad')
+            loss_and_metrics = self.model.training_step(batch, batch_idx)
+            w_step_loss = loss_and_metrics['loss'] \
+                if isinstance(loss_and_metrics, dict) else loss_and_metrics
+            self.manual_backward(w_step_loss)
+            self.call_user_optimizers(w_opt, 'step')
+            return loss_and_metrics
+
+        if source == 'val':
+            # step 2: train ENAS agent
+            x, y = batch
+            arc_opt.zero_grad()
+            self._resample()
+            with torch.no_grad():
+                logits = self.model(x)
+            # get the default metric of self.model
+            if len(self.model.metrics) == 1:
+                _, metric = next(iter(self.model.metrics.items()))
+            else:
+                if 'default' not in self.model.metrics.keys():
+                    raise KeyError('model.metrics should contain a ``default`` key when' \
+                        'there are multiple metrics')
+                metric = self.model.metrics['default']
+
+            reward = metric(logits, y)
+            if self.entropy_weight:
+                reward = reward + self.entropy_weight * self.controller.sample_entropy.item()
+            self.baseline = self.baseline * self.baseline_decay + reward * (1 - self.baseline_decay)
+            rnn_step_loss = self.controller.sample_log_prob * (reward - self.baseline)
+            if self.skip_weight:
+                rnn_step_loss = rnn_step_loss + self.skip_weight * self.controller.sample_skip_penalty
+
+            rnn_step_loss = rnn_step_loss / self.ctrl_steps_aggregate
+            self.manual_backward(rnn_step_loss)
+
+            if (batch_idx + 1) % self.ctrl_steps_aggregate == 0:
+                if self.grad_clip > 0:
+                    nn.utils.clip_grad_norm_(self.controller.parameters(), self.grad_clip)
+                arc_opt.step()
+                arc_opt.zero_grad()
```

Comment:
no return here?

Reply:
This is ENAS agent training phase. What should be returned?